lib/Serializer.ts

```diff
@@ -0,0 +1,120 @@
+	return includeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(document).includes(key)) {
+			serialized[key] = document[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const excludeHandler = (document: ModelType<Document>, excludeRules: string[], serialized: ObjectType): ObjectType => {
+	return excludeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(serialized).includes(key)) {
+			delete serialized[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const validateName = (name: string): void => {
+	if (!name || typeof name !== "string") {
+		throw new Error("Field name is required and should be of type string");
+	}
+};
+
+const validateOptions = (options): void => {
+	if (!options || !(Array.isArray(options) || typeof options === "object")) {
+		throw new Error("Field options is required and should be an object or array");
+	}
+};
+
+const cleanAndValidateDocumentsArray = (documentsArray: ModelType<Document>[]): ModelType<Document>[] => {
+	if (!documentsArray || !Array.isArray(documentsArray)) {
+		throw new Error("documentsArray must be an array of document objects");
+	}
+
+	return documentsArray.filter((doc) => typeof doc.serialize === "function");
+};
+
+export class Serializer {
+	#serializers: any;
+	#defaultSerializer: string;
+
+	constructor() {
+		this.#serializers = {_default: defaultSerializer};
+		this.#defaultSerializer = "_default";
+	}
+
+	add(name: string, options): void {
+		validateName(name);
+		validateOptions(options);
+		this.#serializers[name] = options;
+	}
+
+	setDefault(name: string): void {
+		validateName(name);
+		if (Object.keys(this.#serializers).includes(name)) {
+			this.#defaultSerializer = name;
+		}
+	}
+
+	remove(name: string): void {
+		validateName(name);
+		if (Object.keys(this.#serializers).includes(name)) {
+			delete this.#serializers[name];
+		}
+		if (this.#defaultSerializer === name) {
+			this.#defaultSerializer = "_default";
+		}
+	}
+
+	_serializeMany(documentsArray: ModelType<Document>[] = [], nameOrOptions): ObjectType[] {
+		documentsArray = cleanAndValidateDocumentsArray(documentsArray);
+		return documentsArray.map((doc) => doc.serialize(nameOrOptions));
+	}
+
+	_serialize(document: ModelType<Document>, nameOrOptions = this.#defaultSerializer): ObjectType {
+		const inputType = typeof nameOrOptions;
+		let options;
+
+		if (inputType === "string") {
+			options = this.#serializers[nameOrOptions];
+		} else if (Array.isArray(nameOrOptions) || inputType === "object") {
+			options = nameOrOptions;
+		}
+
+		validateOptions(options);
+		if (Array.isArray(options)) {
+			return includeHandler(document, options, {});
+		}
+
+		let serialized: ObjectType = {};
+		if (options.include) {
+			serialized = includeHandler(document, options.include, serialized);
+		}
+		if (options.exclude) {
+			if (!options.include) {
```

Comment:
In what case would you want to pass in both `exclude` & `include`? I think we should put a limit where it only works with one or the other.

Reply:
So this is potentially future proofing for an additional feature (found in the library I based this work on). Basically they allowed you to use wildcard tokens that you could like include all foreign keys (maybe say `@keys` in the include or exclude arrays that does the appropriate action for all hash+range key values in the document). So if you included all keys but then also wanted to exclude a certain one you could do that. You're right that its somewhat pointless for the current iteration but I think not really an issue to leave as is.

Reply:
@ablankenship10 Something like having include be `friends.**` (to include all nested attributes in `friends`) and exclude be `friends.*.age` (to exclude that one specific property)?

Reply:
I don't think the current code supports nested documents. I wouldn't necessarily handle it like this as you would start having to configure values for serializers of other Models and it gets messy that way. The way I've done it before with the other library is using the modify function, if the current document being serialized contains another "joined" document, to set its value to the serialized.otherDocument = otherDocument.serialize() and run its own serializers configured in its own model.

Reply:
It does support nested objects tho. So you could have an attribute that is `friends.age` or `friends.name` which would be a nested object within the document.